std/hxcoro/ds/ConcurrentCircularBuffer.hx

```diff
@@ -0,0 +1,82 @@
+package hxcoro.ds;
+
+import haxe.ds.Vector;
+import haxe.exceptions.ArgumentException;
+import hxcoro.concurrent.AtomicInt;
+
+/**
+ * Thread safe FIFO circular buffer.
+ * 
+ * This buffer supports at most a single producer and a single consumer at any one time,
+ * the behaviour when multiple produces and consumers act on the buffer is undefined.
+ */
+final class ConcurrentCircularBuffer<T> {
+	final storage : Vector<T>;
+
+	final head : AtomicInt;
+
+	final tail : AtomicInt;
+
+	public function new(capacity : Int) {
+		if (capacity < 1) {
+			throw new ArgumentException("capacity", "Capacity must be greater than zero");
+		}
+
+		// We need +1 since we do a "full" check by comparing the head and tail.
+		storage = new Vector(capacity + 1);
+		head    = new AtomicInt(0);
+		tail    = new AtomicInt(0);
+	}
+
+	/**
+	 * Attempts to add an item to the end of the buffer.
+	 * @param v Item to add.
+	 * @returns `true` if the item was added to the buffer, otherwise `false`.
+	 */
+	public function tryPush(v : T) {
+		final currentTail = tail.load();
+		final nextTail    = increment(currentTail);
+
+		if (nextTail != head.load()) {
+			storage[currentTail] = v;
+			tail.store(nextTail);
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Attempts to remove an item from the beginning of the buffer.
+	 * @param out If this function returns `true` the removed item will be stored in this out object.
+	 * @returns `true` if an item was removed from the buffer, otherwise `false`.
+	 */
+	public function tryPop(out : Out<T>) {
+		final currentHead = head.load();
+		if (currentHead == tail.load()) {
+			return false;
+		}
+
+		// Note : We should probably wipe the previous value here to prevent references being kept to otherwise dead objects.
+		// is it safe to do a `= null` even if the circular buffer is storing, say, ints?
```

Comment:
The way it currently is we don't have actual `Int` storage anyway because we're working with a generic `Vector<T>`. This is a larger problem, but it's good for this specific question because setting the storage to `null` should indeed be safe.

In the long run, all data structures that work on top of `Vector` should become `@:generic`. I wrestled with that for `PagedDeque` as well and it caused too many issues, so we should table that for the time being.